openedx/core/djangoapps/xblock/tests/test_utils.py

```diff
@@ -0,0 +1,135 @@
+        # but the old secret key is in the token keys list.
+        (
+            {
+                "validation_xblock_handler_token_keys": ["baseline_key", ],
+                "validation_secret_key": "new secret key",
+            },
+            True,
+        ),
+        # Test tokens generated with the old key will be valid when there is a new primary
+        # token key at validation time.
+        (
+            {
+                "generation_xblock_handler_token_keys": ["baseline_key", ],
+                "validation_xblock_handler_token_keys": ["now token key", "baseline_key", ],
+                "validation_secret_key": "new secret key",
+            },
+            True,
+        ),
+        # Test token generated with the new key is valid with the new key even if the secret key
+        # changes.
+        (
+            {
+                "generation_xblock_handler_token_keys": ["new token key"],
+                "validation_xblock_handler_token_keys": ["new token key"],
+                "validation_secret_key": "new secret key",  # Ensure that we're not matching with secret key
+            },
+            True,
+        ),
+    ],
+)
+def test_secure_token(param_delta: dict, expected_validation: bool):
+    params = {}
+    params.update(REFERENCE_PARAMS)
+    params.update(param_delta)
+    reference_time = params['reference_time']
+
+    with override_settings(
+        SECRET_KEY=params["generation_secret_key"],
+        XBLOCK_HANDLER_TOKEN_KEYS=params["generation_xblock_handler_token_keys"],
+    ):
+        with freeze_time(reference_time):
+            token = get_secure_token_for_xblock_handler(
+                params["generation_user_id"], params["generation_block_key_str"]
+            )
+
+    with override_settings(
+        SECRET_KEY=params["validation_secret_key"],
+        XBLOCK_HANDLER_TOKEN_KEYS=params["validation_xblock_handler_token_keys"],
+    ):
+        with freeze_time(reference_time + datetime.timedelta(seconds=params["validation_time_delta_s"])):
+            assert (
+                validate_secure_token_for_xblock_handler(
+                    params["validation_user_id"], params["validation_block_key_str"], token
+                )
+                == expected_validation
+            )
+
+def test_private_get_secure_token_for_xblock_handler():
+    """
+    Confirms function behaviour has not changed and is giving the same token for same inputs
+    If these tokens change, this will invalidate some or all of the tokens handed out to learners
+    """
+    with freeze_time(datetime.datetime(2021, 1, 27, 0, 0, 0, tzinfo=datetime.timezone.utc)):
+        token_now = _get_secure_token_for_xblock_handler("12345", "some_block_key", 0, "some_hashing_key")
+        token_previous = _get_secure_token_for_xblock_handler("12345", "some_block_key", -1, "some_hashing_key")
+        assert token_now == "de8399a51fef6aa7584a"
+        assert token_previous == "f3b5d92b8ca2934009e4"
```

Comment:
We confirmed that when this test was run before and after the fix, only `token_previous` changed.